File: app/src/main/java/com/biginsect/easyhub/mvp/BaseMvpFragment.java
package com.biginsect.easyhub.mvp;

import android.app.Activity;
import android.content.Context;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.view.View;


/**
 * @author big insect
 * @date 2018/9/3.
 */

public abstract class BaseMvpFragment<V extends MvpView, P extends MvpPresenter<V>>
        extends Fragment
        implements MvpView, MvpDelegateCallback<V, P> {

    protected FragmentMvpDelegate mvpDelegate;
    protected P presenter;

    @SuppressWarnings("unchecked")
    protected FragmentMvpDelegate<V, P> getMvpDelegate() {
        if (null == mvpDelegate) {
            mvpDelegate = new FragmentMvpDelegateImpl(this, this, true, true);
        }

        return mvpDelegate;
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        getMvpDelegate().onViewCreated(view, savedInstanceState);
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        getMvpDelegate().onCreate(savedInstanceState);
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        getMvpDelegate().onDestroy();
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        getMvpDelegate().onSaveInstanceState(outState);
    }

    @Override
    public void onPause() {
        super.onPause();
        getMvpDelegate().onPause();
    }

    @Override
    public void onResume() {
        super.onResume();
        getMvpDelegate().onResume();
    }

    @Override
    public void onStart() {
        super.onStart();
        getMvpDelegate().onStart();
    }

    @Override
    public void onStop() {
        super.onStop();
        getMvpDelegate().onStop();
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        getMvpDelegate().onActivityCreated(savedInstanceState);
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        getMvpDelegate().onDestroyView();
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        getMvpDelegate().onAttach((Activity) context);
    }

    @Override
    public void onDetach() {
        super.onDetach();
        getMvpDelegate().onDetach();
    }

    @SuppressWarnings("unchecked")
    @Nullable
    @Override
    public V getMvpView() {
        return (V) this;
    }

    @Nullable
    @Override
    public P getPresenter() {
        return presenter;
    }

    @Override
    public void setPresenter(P presenter) {
        this.presenter = presenter;
    }

    /**
     * create presenter
     *
     * @return target
     */
    @NonNull
    @Override
    public abstract P createPresenter();
}

File: app/src/main/java/com/biginsect/easyhub/mvp/BaseMvpFragment.kt
package com.biginsect.easyhub.mvp

import android.app.Activity
import android.content.Context
import android.os.Bundle
import android.support.v4.app.Fragment
import android.view.View

/**
 * @author big insect
 */
abstract class BaseMvpFragment<V : MvpView, P : MvpPresenter<V>>: Fragment()
        , MvpDelegateCallback<V, P>, MvpView {
    private var mvpDelegate: FragmentMvpDelegate<V, P>? = null


    protected fun getDelegate():FragmentMvpDelegate<V, P>{
        if (null == mvpDelegate){
            mvpDelegate = FragmentMvpDelegateImpl(this, this, true, true)
        }

        return mvpDelegate as FragmentMvpDelegate<V, P>
    }

//    protected var mPresenter : P? = null

    @Suppress("UNCHECKED_CAST")
    override fun getMvpView(): V? {
        return this as V
    }
    /*override fun setPresenter(presenter: P?) {
        this.mPresenter = presenter
    }

    override fun getPresenter(): P? {
        return mPresenter
    }*/

    abstract override fun createPresenter(): P?

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        getDelegate().onViewCreated(view, savedInstanceState)
    }

    override fun onDestroyView() {
        super.onDestroyView()
        getDelegate().onDestroyView()
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        getDelegate().onCreate(savedInstanceState)
    }

    override fun onDestroy() {
        super.onDestroy()
        getDelegate().onDestroy()
    }

    override fun onPause() {
        super.onPause()
        getDelegate().onPause()
    }

    override fun onResume() {
        super.onResume()
        getDelegate().onResume()
    }

    override fun onStart() {
        super.onStart()
        getDelegate().onStart()
    }

    override fun onStop() {
        super.onStop()
        getDelegate().onStop()
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        getDelegate().onActivityCreated(savedInstanceState)
    }

    override fun onAttach(context: Context?) {
        super.onAttach(context)
        getDelegate().onAttach(context as Activity)
    }

    override fun onDetach() {
        super.onDetach()
        getDelegate().onDetach()
    }

    override fun onSaveInstanceState(outState: Bundle?) {
        super.onSaveInstanceState(outState)
        getDelegate().onSaveInstanceState(outState)
    }
}